File: app/src/main/kotlin/com/inkframe/studio/vector/VectorEngine.kt
package com.inkframe.studio.vector

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.ceil
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.round
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Pure Kotlin vector geometry core for InkFrame.
 *
 * This is renderer-agnostic: no Android Canvas, no Compose, no WebView. It gives
 * the app a native vector foundation inspired by professional vector workflows:
 * editable anchors, cubic Bézier paths, simplification, snapping, symmetry, and
 * stroke-outline planning.
 */
object VectorEngine {
    const val VERSION: String = "v0.1.0-vector-path-core"

    private const val EPSILON = 1e-5f

    enum class AnchorKind { Corner, Smooth, Symmetric }
    enum class PathFillRule { NonZero, EvenOdd }
    enum class SnapMode { None, Grid, Angle, GridAndAngle }
    enum class SymmetryMode { None, Horizontal, Vertical, Quad }

    data class Vec2(val x: Float, val y: Float) {
        operator fun plus(other: Vec2): Vec2 = Vec2(x + other.x, y + other.y)
        operator fun minus(other: Vec2): Vec2 = Vec2(x - other.x, y - other.y)
        operator fun times(scale: Float): Vec2 = Vec2(x * scale, y * scale)
        operator fun div(scale: Float): Vec2 = Vec2(x / scale, y / scale)
        fun dot(other: Vec2): Float = x * other.x + y * other.y
        fun cross(other: Vec2): Float = x * other.y - y * other.x
        fun length(): Float = sqrt(x * x + y * y)
        fun distanceTo(other: Vec2): Float = (this - other).length()
        fun normalized(): Vec2 {
            val len = length()
            return if (len < EPSILON) Zero else this / len
        }
        fun perpendicularLeft(): Vec2 = Vec2(-y, x)
        fun lerp(to: Vec2, t: Float): Vec2 = this + (to - this) * t.coerceIn(0f, 1f)
        fun mirrorX(centerX: Float): Vec2 = Vec2(centerX * 2f - x, y)
        fun mirrorY(centerY: Float): Vec2 = Vec2(x, centerY * 2f - y)

        companion object {
            val Zero = Vec2(0f, 0f)
        }
    }

    data class Bounds(val minX: Float, val minY: Float, val maxX: Float, val maxY: Float) {
        val width: Float get() = maxX - minX
        val height: Float get() = maxY - minY
        val center: Vec2 get() = Vec2((minX + maxX) / 2f, (minY + maxY) / 2f)
        fun include(point: Vec2): Bounds = Bounds(
            min(minX, point.x),
            min(minY, point.y),
            max(maxX, point.x),
            max(maxY, point.y),
        )

        companion object {
            fun from(points: List<Vec2>): Bounds {
                if (points.isEmpty()) return Bounds(0f, 0f, 0f, 0f)
                var out = Bounds(points.first().x, points.first().y, points.first().x, points.first().y)
                points.drop(1).forEach { out = out.include(it) }
                return out
            }
        }
    }

    data class AnchorNode(
        val point: Vec2,
        val handleIn: Vec2? = null,
        val handleOut: Vec2? = null,
        val kind: AnchorKind = AnchorKind.Smooth,
    )

    data class CubicBezier(
        val start: Vec2,
        val control1: Vec2,
        val control2: Vec2,
        val end: Vec2,
    )

    data class VectorStyle(
        val strokeColor: Int = 0xFF111111.toInt(),
        val fillColor: Int? = null,
        val strokeWidth: Float = 4f,
        val opacity: Float = 1f,
        val fillRule: PathFillRule = PathFillRule.NonZero,
    ) {
        fun sanitized(): VectorStyle = copy(
            strokeWidth = strokeWidth.coerceIn(0f, 2048f),
            opacity = opacity.coerceIn(0f, 1f),
        )
    }

    data class VectorPath(
        val anchors: List<AnchorNode>,
        val closed: Boolean = false,
        val style: VectorStyle = VectorStyle(),
        val name: String = "Path",
    ) {
        val isRenderable: Boolean get() = anchors.size >= 2
    }

    data class SnapConfig(
        val mode: SnapMode = SnapMode.None,
        val gridSize: Float = 16f,
        val angleStepDegrees: Float = 15f,
        val origin: Vec2 = Vec2.Zero,
    )

    data class StrokeOutline(val left: List<Vec2>, val right: List<Vec2>) {
        val polygon: List<Vec2> get() = left + right.asReversed()
    }

    data class VectorPlan(
        val rawPoints: List<Vec2>,
        val simplifiedPoints: List<Vec2>,
        val anchors: List<AnchorNode>,
        val cubics: List<CubicBezier>,
        val samples: List<Vec2>,
        val outline: StrokeOutline,
        val bounds: Bounds,
    )

    fun planVectorStroke(
        rawPoints: List<Vec2>,
        simplificationTolerance: Float = 1.25f,
        closed: Boolean = false,
        style: VectorStyle = VectorStyle(),
        sampleStep: Float = 0.08f,
    ): VectorPlan {
        val clean = rawPoints.filterFinite().dedupeNear()
        val simplified = simplify(clean, simplificationTolerance)
        val cubics = catmullRomToCubics(simplified, closed = closed)
        val anchors = anchorsFromCubics(cubics, fallback = simplified)
        val samples = sampleCubics(cubics, sampleStep).ifEmpty { simplified }
        val outline = outlinePolyline(samples, style.sanitized().strokeWidth)
        val bounds = Bounds.from(samples + outline.polygon)
        return VectorPlan(clean, simplified, anchors, cubics, samples, outline, bounds)
    }

    fun simplify(points: List<Vec2>, tolerance: Float): List<Vec2> {
        if (points.size <= 2 || tolerance <= 0f) return points
        return rdp(points, tolerance.coerceAtLeast(0f))
    }

    fun catmullRomToCubics(points: List<Vec2>, closed: Boolean = false, tension: Float = 1f): List<CubicBezier> {
        if (points.size < 2) return emptyList()
        val t = tension.coerceIn(0f, 2f)
        val out = mutableListOf<CubicBezier>()
        val count = if (closed) points.size else points.size - 1
        for (i in 0 until count) {
            val p0 = points.getLooped(i - 1, closed) ?: points[i]
            val p1 = points.getLooped(i, closed) ?: points[i]
            val p2 = points.getLooped(i + 1, closed) ?: points[min(i + 1, points.lastIndex)]
            val p3 = points.getLooped(i + 2, closed) ?: p2
            val c1 = p1 + (p2 - p0) * (t / 6f)
            val c2 = p2 - (p3 - p1) * (t / 6f)
            out += CubicBezier(p1, c1, c2, p2)
        }
        return out
    }

    fun anchorsFromCubics(cubics: List<CubicBezier>, fallback: List<Vec2> = emptyList()): List<AnchorNode> {
        if (cubics.isEmpty()) return fallback.map { AnchorNode(it, kind = AnchorKind.Corner) }
        val anchors = mutableListOf<AnchorNode>()
        anchors += AnchorNode(cubics.first().start, handleOut = cubics.first().control1, kind = AnchorKind.Smooth)
        cubics.forEachIndexed { index, cubic ->
            val next = cubics.getOrNull(index + 1)
            anchors += AnchorNode(
                point = cubic.end,
                handleIn = cubic.control2,
                handleOut = next?.control1,
                kind = AnchorKind.Smooth,
            )
        }
        return anchors
    }

    fun sampleCubic(cubic: CubicBezier, t: Float): Vec2 {
        val u = t.coerceIn(0f, 1f)
        val inv = 1f - u
        val a = inv * inv * inv
        val b = 3f * inv * inv * u
        val c = 3f * inv * u * u
        val d = u * u * u
        return cubic.start * a + cubic.control1 * b + cubic.control2 * c + cubic.end * d
    }

    fun sampleCubics(cubics: List<CubicBezier>, step: Float = 0.08f): List<Vec2> {
        if (cubics.isEmpty()) return emptyList()
        val safeStep = step.coerceIn(0.01f, 1f)
        val perSegment = ceil(1f / safeStep).toInt().coerceIn(1, 512)
        val out = mutableListOf<Vec2>()
        cubics.forEachIndexed { index, cubic ->
            if (index == 0) out += cubic.start
            for (i in 1..perSegment) out += sampleCubic(cubic, i.toFloat() / perSegment)
        }
        return out.dedupeNear()
    }

    fun outlinePolyline(points: List<Vec2>, width: Float): StrokeOutline {
        if (points.isEmpty()) return StrokeOutline(emptyList(), emptyList())
        val half = max(0f, width) / 2f
        if (points.size == 1 || half <= EPSILON) return StrokeOutline(points, points)
        val left = mutableListOf<Vec2>()
        val right = mutableListOf<Vec2>()
        points.forEachIndexed { i, p ->
            val prev = points.getOrNull(i - 1) ?: p
            val next = points.getOrNull(i + 1) ?: p
            val tangent = (next - prev).normalized()
            val normal = tangent.perpendicularLeft()
            left += p + normal * half
            right += p - normal * half
        }
        return StrokeOutline(left, right)
    }

    fun snapPoint(point: Vec2, previous: Vec2? = null, config: SnapConfig): Vec2 {
        var out = point
        if (config.mode == SnapMode.Grid || config.mode == SnapMode.GridAndAngle) {
            val grid = max(EPSILON, config.gridSize)
            out = Vec2(
                config.origin.x + round((out.x - config.origin.x) / grid) * grid,
                config.origin.y + round((out.y - config.origin.y) / grid) * grid,
            )
        }
        if ((config.mode == SnapMode.Angle || config.mode == SnapMode.GridAndAngle) && previous != null) {
            val delta = out - previous
            val len = delta.length()
            if (len > EPSILON) {
                val step = (config.angleStepDegrees.coerceIn(1f, 90f) * PI.toFloat()) / 180f
                val angle = atan2(delta.y, delta.x)
                val snapped = round(angle / step) * step
                out = previous + Vec2(cos(snapped), sin(snapped)) * len
            }
        }
        return out
    }

    fun symmetryCopies(points: List<Vec2>, mode: SymmetryMode, center: Vec2): List<List<Vec2>> = when (mode) {
        SymmetryMode.None -> listOf(points)
        SymmetryMode.Horizontal -> listOf(points, points.map { it.mirrorY(center.y) })
        SymmetryMode.Vertical -> listOf(points, points.map { it.mirrorX(center.x) })
        SymmetryMode.Quad -> listOf(
            points,
            points.map { it.mirrorX(center.x) },
            points.map { it.mirrorY(center.y) },
            points.map { it.mirrorX(center.x).mirrorY(center.y) },
        )
    }

    fun svgPathData(cubics: List<CubicBezier>, closed: Boolean = false): String {
        if (cubics.isEmpty()) return ""
        val b = StringBuilder()
        b.append("M ").append(fmt(cubics.first().start.x)).append(' ').append(fmt(cubics.first().start.y))
        cubics.forEach { c ->
            b.append(" C ")
                .append(fmt(c.control1.x)).append(' ').append(fmt(c.control1.y)).append(", ")
                .append(fmt(c.control2.x)).append(' ').append(fmt(c.control2.y)).append(", ")
                .append(fmt(c.end.x)).append(' ').append(fmt(c.end.y))
        }
        if (closed) b.append(" Z")
        return b.toString()
    }

    private fun List<Vec2>.filterFinite(): List<Vec2> = filter { it.x.isFinite() && it.y.isFinite() }

    private fun List<Vec2>.dedupeNear(threshold: Float = 0.05f): List<Vec2> {
        if (isEmpty()) return this
        val out = mutableListOf(first())
        drop(1).forEach { if (it.distanceTo(out.last()) > threshold) out += it }
        return out
    }

    private fun rdp(points: List<Vec2>, tolerance: Float): List<Vec2> {
        if (points.size <= 2) return points
        var maxDistance = -1f
        var index = -1
        val start = points.first()
        val end = points.last()
        for (i in 1 until points.lastIndex) {
            val distance = perpendicularDistance(points[i], start, end)
            if (distance > maxDistance) {
                maxDistance = distance
                index = i
            }
        }
        return if (maxDistance > tolerance && index > 0) {
            val left = rdp(points.subList(0, index + 1), tolerance)
            val right = rdp(points.subList(index, points.size), tolerance)
            left.dropLast(1) + right
        } else {
            listOf(start, end)
        }
    }

    private fun perpendicularDistance(p: Vec2, a: Vec2, b: Vec2): Float {
        val ab = b - a
        val len = ab.length()
        if (len < EPSILON) return p.distanceTo(a)
        return abs((p - a).cross(ab)) / len
    }

    private fun List<Vec2>.getLooped(index: Int, closed: Boolean): Vec2? {
        if (isEmpty()) return null
        if (!closed && index !in indices) return null
        val wrapped = ((index % size) + size) % size
        return this[wrapped]
    }

    private fun fmt(value: Float): String {
        val rounded = floor(value * 1000f + 0.5f) / 1000f
        return rounded.toString().trimEnd('0').trimEnd('.')
    }
}

File: web/vector-engine.js
// InkFrame — Vector Path Engine Core
// -----------------------------------------------------------------------------
// Portable vector geometry primitives for editable strokes and future SVG/native
// rendering. Mirrors the Kotlin VectorEngine shape: points, cubic Béziers,
// simplification, Catmull-Rom conversion, snapping, symmetry, outlines, bounds.
'use strict';

(function installInkFrameVectorEngine(root){
  const VERSION = 'v0.1.0-vector-path-core';
  const EPSILON = 1e-5;
  const clamp = (min, value, max) => Math.max(min, Math.min(max, Number(value) || 0));
  const finite = value => Number.isFinite(Number(value));

  function vec(x, y){ return Object.freeze({ x: Number(x) || 0, y: Number(y) || 0 }); }
  function add(a, b){ return vec(a.x + b.x, a.y + b.y); }
  function sub(a, b){ return vec(a.x - b.x, a.y - b.y); }
  function mul(a, s){ return vec(a.x * s, a.y * s); }
  function div(a, s){ return vec(a.x / s, a.y / s); }
  function length(a){ return Math.hypot(a.x, a.y); }
  function distance(a, b){ return length(sub(a, b)); }
  function normalized(a){ const len = length(a); return len < EPSILON ? vec(0, 0) : div(a, len); }
  function cross(a, b){ return a.x * b.y - a.y * b.x; }
  function leftNormal(a){ return vec(-a.y, a.x); }
  function lerp(a, b, t){ return add(a, mul(sub(b, a), clamp(0, t, 1))); }
  function mirrorX(p, cx){ return vec(cx * 2 - p.x, p.y); }
  function mirrorY(p, cy){ return vec(p.x, cy * 2 - p.y); }

  function bounds(points){
    if (!points || !points.length) return Object.freeze({ minX:0, minY:0, maxX:0, maxY:0, width:0, height:0, center:vec(0,0) });
    let minX = points[0].x, minY = points[0].y, maxX = points[0].x, maxY = points[0].y;
    points.forEach(p => { minX = Math.min(minX, p.x); minY = Math.min(minY, p.y); maxX = Math.max(maxX, p.x); maxY = Math.max(maxY, p.y); });
    return Object.freeze({ minX, minY, maxX, maxY, width:maxX-minX, height:maxY-minY, center:vec((minX+maxX)/2, (minY+maxY)/2) });
  }

  function cleanPoints(points){
    const out = [];
    (points || []).forEach(raw => {
      if (!raw || !finite(raw.x) || !finite(raw.y)) return;
      const p = vec(raw.x, raw.y);
      if (!out.length || distance(p, out[out.length - 1]) > 0.05) out.push(p);
    });
    return out;
  }

  function perpendicularDistance(p, a, b){
    const ab = sub(b, a);
    const len = length(ab);
    if (len < EPSILON) return distance(p, a);
    return Math.abs(cross(sub(p, a), ab)) / len;
  }

  function simplify(points, tolerance){
    const pts = cleanPoints(points);
    const tol = Math.max(0, Number(tolerance) || 0);
    if (pts.length <= 2 || tol <= 0) return pts;
    function rdp(slice){
      if (slice.length <= 2) return slice;
      const start = slice[0], end = slice[slice.length - 1];
      let maxDistance = -1, index = -1;
      for (let i = 1; i < slice.length - 1; i++) {
        const d = perpendicularDistance(slice[i], start, end);
        if (d > maxDistance) { maxDistance = d; index = i; }
      }
      if (maxDistance > tol && index > 0) return rdp(slice.slice(0, index + 1)).slice(0, -1).concat(rdp(slice.slice(index)));
      return [start, end];
    }
    return rdp(pts);
  }

  function getLooped(points, index, closed){
    if (!points.length) return null;
    if (!closed && (index < 0 || index >= points.length)) return null;
    return points[((index % points.length) + points.length) % points.length];
  }

  function catmullRomToCubics(points, opts){
    const pts = cleanPoints(points);
    const closed = !!(opts && opts.closed);
    const tension = clamp(0, opts && opts.tension == null ? 1 : opts.tension, 2);
    if (pts.length < 2) return [];
    const out = [];
    const count = closed ? pts.length : pts.length - 1;
    for (let i = 0; i < count; i++) {
      const p0 = getLooped(pts, i - 1, closed) || pts[i];
      const p1 = getLooped(pts, i, closed) || pts[i];
      const p2 = getLooped(pts, i + 1, closed) || pts[Math.min(i + 1, pts.length - 1)];
      const p3 = getLooped(pts, i + 2, closed) || p2;
      out.push(Object.freeze({
        start: p1,
        control1: add(p1, mul(sub(p2, p0), tension / 6)),
        control2: sub(p2, mul(sub(p3, p1), tension / 6)),
        end: p2,
      }));
    }
    return out;
  }

  function anchorsFromCubics(cubics, fallback){
    if (!cubics || !cubics.length) return (fallback || []).map(p => Object.freeze({ point:p, handleIn:null, handleOut:null, kind:'corner' }));
    const anchors = [Object.freeze({ point:cubics[0].start, handleIn:null, handleOut:cubics[0].control1, kind:'smooth' })];
    cubics.forEach((c, i) => anchors.push(Object.freeze({ point:c.end, handleIn:c.control2, handleOut:cubics[i + 1] ? cubics[i + 1].control1 : null, kind:'smooth' })));
    return anchors;
  }

  function sampleCubic(c, t){
    const u = clamp(0, t, 1);
    const inv = 1 - u;
    return add(add(mul(c.start, inv*inv*inv), mul(c.control1, 3*inv*inv*u)), add(mul(c.control2, 3*inv*u*u), mul(c.end, u*u*u)));
  }

  function sampleCubics(cubics, step){
    const safeStep = clamp(0.01, step == null ? 0.08 : step, 1);
    const perSegment = Math.max(1, Math.min(512, Math.ceil(1 / safeStep)));
    const out = [];
    (cubics || []).forEach((c, index) => {
      if (index === 0) out.push(c.start);
      for (let i = 1; i <= perSegment; i++) out.push(sampleCubic(c, i / perSegment));
    });
    return cleanPoints(out);
  }

  function outlinePolyline(points, width){
    const pts = cleanPoints(points);
    const half = Math.max(0, Number(width) || 0) / 2;
    if (!pts.length) return Object.freeze({ left:[], right:[], polygon:[] });
    if (pts.length === 1 || half <= EPSILON) return Object.freeze({ left:pts, right:pts, polygon:pts.concat(pts.slice().reverse()) });
    const left = [], right = [];
    pts.forEach((p, i) => {
      const prev = pts[i - 1] || p;
      const next = pts[i + 1] || p;
      const normal = leftNormal(normalized(sub(next, prev)));
      left.push(add(p, mul(normal, half)));
      right.push(sub(p, mul(normal, half)));
    });
    return Object.freeze({ left, right, polygon:left.concat(right.slice().reverse()) });
  }

  function snapPoint(point, previous, config){
    const cfg = { mode:'none', gridSize:16, angleStepDegrees:15, origin:vec(0,0), ...(config || {}) };
    let out = vec(point.x, point.y);
    if (cfg.mode === 'grid' || cfg.mode === 'gridAndAngle') {
      const grid = Math.max(EPSILON, Number(cfg.gridSize) || 16);
      out = vec(cfg.origin.x + Math.round((out.x - cfg.origin.x) / grid) * grid, cfg.origin.y + Math.round((out.y - cfg.origin.y) / grid) * grid);
    }
    if ((cfg.mode === 'angle' || cfg.mode === 'gridAndAngle') && previous) {
      const delta = sub(out, previous);
      const len = length(delta);
      if (len > EPSILON) {
        const step = clamp(1, cfg.angleStepDegrees || 15, 90) * Math.PI / 180;
        const snapped = Math.round(Math.atan2(delta.y, delta.x) / step) * step;
        out = add(previous, mul(vec(Math.cos(snapped), Math.sin(snapped)), len));
      }
    }
    return out;
  }

  function symmetryCopies(points, mode, center){
    const pts = cleanPoints(points);
    const c = center || vec(0,0);
    if (mode === 'horizontal') return [pts, pts.map(p => mirrorY(p, c.y))];
    if (mode === 'vertical') return [pts, pts.map(p => mirrorX(p, c.x))];
    if (mode === 'quad') return [pts, pts.map(p => mirrorX(p, c.x)), pts.map(p => mirrorY(p, c.y)), pts.map(p => mirrorY(mirrorX(p, c.x), c.y))];
    return [pts];
  }

  function fmt(value){
    return String(Math.round(value * 1000) / 1000).replace(/\.0+$/, '').replace(/(\.\d*?)0+$/, '$1');
  }

  function svgPathData(cubics, closed){
    if (!cubics || !cubics.length) return '';
    let d = `M ${fmt(cubics[0].start.x)} ${fmt(cubics[0].start.y)}`;
    cubics.forEach(c => { d += ` C ${fmt(c.control1.x)} ${fmt(c.control1.y)}, ${fmt(c.control2.x)} ${fmt(c.control2.y)}, ${fmt(c.end.x)} ${fmt(c.end.y)}`; });
    if (closed) d += ' Z';
    return d;
  }

  function planVectorStroke(rawPoints, options){
    const opts = { simplificationTolerance:1.25, closed:false, strokeWidth:4, sampleStep:0.08, ...(options || {}) };
    const raw = cleanPoints(rawPoints);
    const simplified = simplify(raw, opts.simplificationTolerance);
    const cubics = catmullRomToCubics(simplified, { closed:opts.closed, tension:opts.tension == null ? 1 : opts.tension });
    const anchors = anchorsFromCubics(cubics, simplified);
    const samples = sampleCubics(cubics, opts.sampleStep);
    const renderSamples = samples.length ? samples : simplified;
    const outline = outlinePolyline(renderSamples, opts.strokeWidth);
    const b = bounds(renderSamples.concat(outline.polygon));
    return Object.freeze({ rawPoints:raw, simplifiedPoints:simplified, anchors, cubics, samples:renderSamples, outline, bounds:b, svgPathData:svgPathData(cubics, opts.closed) });
  }

  const api = Object.freeze({
    VERSION,
    vec,
    add,
    sub,
    mul,
    length,
    distance,
    normalized,
    bounds,
    simplify,
    catmullRomToCubics,
    anchorsFromCubics,
    sampleCubic,
    sampleCubics,
    outlinePolyline,
    snapPoint,
    symmetryCopies,
    svgPathData,
    planVectorStroke,
  });

  if (root && typeof root === 'object') root.InkFrameVectorEngine = api;
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
})(typeof window !== 'undefined' ? window : globalThis);
